### PR TITLE
Display Dynamic Cards in the Dashboard

### LINKED ISSUE #22150: 🍎 [Dynamic Cards] Update Endpoint Response
### Endpoint
`sites/{SITE_ID}/dashboard/cards-data`

### The Full Response: 
```
{
  "todays_stats": {...},
  "posts": {...},
  "pages": [...],
  "activity": {...},
  "dynamic": [Card, Card, ...]
}
```

### The Card Object:
```
{
  "id": "year_in_review_2023",
  "title": "News",
  "remote_feature_flag": "dynamic_dashboard_cards",
  "featured_image": "https://path/to/image",
  "url": "https://wordpress.com",
  "action": "Call to action",
  "order": "top",
  "rows": [
    {
      "icon": "https://path/to/image",
      "title": "xyz",
      "description": "xyz"
    },
    {
      "icon": "https://path/to/image",
      "title": "xyz",
      "description": "xyz"
    }
  ]
}
```
order will have 3 possible values: “top”, “middle” and “bottom”.

### LINKED ISSUE #22154: 🍎 [Dynamic Cards] Ordering logic
Order type `top` cards will be sorted to the beginning of the card list.
Order type `bottom` cards will be sorted to the end of the card list.

### PR DESCRIPTION
**Pull request by salimbraksa**
Part of #22153
Fixes #22154
Fixes #22150

## Description

This PR displays dynamic cards in the dashboard.

| Light Mode | Dark Mode |
| ----------- | ------------ |
|   |  |

## Test Instructions

1. Install Jetpack and login.
2. **Expect** `New: Domains Management` is not shown.
3. Enable `Dynamic Dashboard Cards` Feature Flag.
4. Navigate to My Site.
5. **Expect** `New: Domains Management` card at the top.

## Regression Notes
1. Potential unintended areas of impact
The service responsible for fetching dashboard cards was refactored. However, the changes include only adding the dynamic cards to the existing list of cards. So technically the existing cards should work as expected. To be safe, it would be better to verify the existing cards are shown.

7. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing.

8. What automated tests I added (or what prevented me from doing so)
None.

## PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/BlogDashboardDynamicCardCell.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+//                    UIMenu(
+//                        options: .displayInline,
+//                        children: [
+//                            BlogDashboardHelpers.makeHideCardAction(for: .googleDomains, blog: blog)
+//                        ]
+//                    )
+//                ],
+//            card: .googleDomains
+//        )
+    }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/BlogDashboardDynamicCardCell.swift:81]: Dismissing dashboard cards is implemented in a separate [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/22229).

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/BlogDashboardCardConfigurable.swift [hunk 2]
```diff
@@ -1,10 +1,29 @@
 import Foundation
 
 protocol BlogDashboardCardConfigurable {
+    func configure(blog: Blog, viewController: BlogDashboardViewController?, model: DashboardCardModel)
     func configure(blog: Blog, viewController: BlogDashboardViewController?, apiResponse: BlogDashboardRemoteEntity?)
+    func configure(blog: Blog, viewController: BlogDashboardViewController?, model: DashboardDynamicCardModel)
     var row: Int { get set }
 }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/BlogDashboardCardConfigurable.swift:8]: Only one of these methods should be implemented.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift [hunk 3]
```diff
@@ -97,27 +97,83 @@ private extension BlogDashboardService {
...
                 return nil
             }
+            return self.dashboardCardModel(
+                from: card,
+                entity: entity,
+                blog: blog,
+                dotComID: dotComID,
+                personalizationService: personalizationService
+            )
+        }
```
Comment (salimbraksa) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:113]: The existing logic was refactored to this `dashboardCardModel` method without any changes to the logic itself.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift [hunk 4]
```diff
@@ -150,9 +153,22 @@ enum DashboardCard: String, CaseIterable {
...
             return FeatureFlag.domainFocus.enabled && isJetpack
+        case .dynamic:
+            return false
         }
     }
 
+    func shouldShow(
+        for blog: Blog,
+        dynamicCardPayload: DashboardDynamicCardModel.Payload,
+        isJetpack: Bool = AppConfiguration.isJetpack
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift:164]: Do we need to pass `isJetpack` in? I assume we'll always pass in `AppConfiguration.isJetpack`. Is it for testability?

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift [hunk 5]
```diff
@@ -97,27 +97,83 @@ private extension BlogDashboardService {
...
 
-            guard card.shouldShow(
-                for: blog,
-                apiResponse: entity,
-                isJetpack: isJetpack,
-                isDotComAvailable: isDotComAvailable,
-                shouldShowJetpackFeatures: shouldShowJetpackFeatures
-            ) else {
-                return nil
+        // Maps dynamic cards to `DashboardCardModel`.k
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:115]: `.k` is a typo?

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift [hunk 6]
```diff
@@ -97,27 +97,83 @@ private extension BlogDashboardService {
...
+            } else {
+                allCards = allCards + bottomCards
             }
+        }
+
+        // Add "empty" card if the list of cards is empty.
+        if allCards.isEmpty || allCards.map(\.cardType) == [.personalize] {
+            let model = DashboardCardModel.normal(.init(cardType: .empty, dotComID: dotComID))
+            allCards.insert(model, at: 0)
+        }
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:139]: Would it be possible to unit test this in isolation? Perhaps create a static function if testing it with the current setup requires too much refactor?

Unit testing this kind of logic could pay in dividends .

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/DashboardServiceRemoteFake.swift [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import Foundation
+import WordPressKit
+
+// TODO: Delete this once dynamic cards are implemented on the Backend
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/DashboardServiceRemoteFake.swift]: Nice helper for development!

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift (hunk 4 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift:164]: @alpavanoglu  That's correct, static methods or properties called within a function implementation are difficult to mock. And I also inspired from the other `shouldShow` method in `DashboardCard` enum.

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift (hunk 5 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:115]: Yes sir!

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift (hunk 6 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:139]: I agree @alpavanoglu, I'll look into unit testing that logic.

Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:139]: @alpavanoglu Resolved in c0f54f59c0169d18c0549871b9969dca1d335310

File: WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift (hunk 5 above)
Comment (salimbraksa) [reply WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Service/BlogDashboardService.swift:115]: @alpavanoglu This was resolved in this [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/22229)

**Comment by alpavanoglu** [comment]
🧪 TEST RESULTS 🧪 

✅ The card appears when the flag is enabled.
✅ Card appears at the top
✅ Dark mode is supported correctly.

**Review by alpavanoglu — APPROVED** [review]

**Pull request merged**